### DISCUSSION TITLE
How to correctly remove properties from a Zowe config file

### DISCUSSION THREAD

**Discussion by juleskreutzer** [zowe/zowe-cli#1561] [Q&A] [answered]
Hello,

In our current zowe implementation, we have several PowerShell scripts that are calling Zowe CLI command under water. These powershell scripts need to be executed from a specific folder, say `%APPDATA%/code4z`, because we also have a `zowe.config.json` file in this folder.

For the current implementation, we use the username and password for authentication to the APIML.

Besides our current PowerShell based implementation, we're developing a new VSCode extension. This extension is also built on top of the Zowe CLI package.

The extension has the following requirements among others:
- Import the previously used `zowe.config.json` file located under `%APPDATA%/code4z`
- Remove/logout from APIML on user request.

**Import the previously used zowe.config.json**
We have decided that we want to import the `%APPDATA%/code4z/zowe.config.json` into the `.zowe` folder and overwrite any existing config file. We currently have the logic in place to copy the file. A part of the file is listed below.
```
{
    ...
            "type": "base",
            "properties": {
                "host": "xxxx",
                "port": xxxx,
                "tokenType": "apimlAuthenticationToken",
                "rejectUnauthorized": xxxx,
                "account": "xxxx",
                "characterSet": "xxxx",
                "codePage": "xxxx"
            },
            "secure": [
                "user",
                "password",
                "tokenValue"
            ]
        }
    ....
```
As you can see, the profile currently has the secure properties `user`, `password` and `tokenValue` and the `tokenType` set to the value `apimlAuthenticationToken`.

In the above example, the `user` and `password` secure properties are not needed when we use the apimlAuthenticationToken. So we want to remove them. Is there functionality provided by the CLI to do so? We can read the config json in and remove the objects, but this doesn't feel like the best solution.

**Remove/logout from APIML on user request**
For security reasons, a user should have the possiblity to logout from our extension. Under water we call the `apimlLogout` function of the Zowe CLI to invalidate the token. Looking at the config file, I can still see the `tokenType` and `tokenValue` in the config file.

We have the following logic to logout the user:
```typescript
if (token === '') {
    throw new GenericError('Unable to retrieve token for logout. Probably it has already been deleted', MessageLocation.CORE);
}

// Create update session
const updSession = new zowe.imperative.Session({
    hostname: MAINFRAME_HOSTNAME,
    port: MAINFRAME_PORT,
    rejectUnauthorized: MAINFRAME_REJECT_UNAUTHORIZED,
    tokenType: MAINFRAME_TOKEN_TYPE,
    type: zowe.imperative.SessConstants.AUTH_TYPE_TOKEN,
    tokenValue: token
});

await zowe.Logout.apimlLogout(updSession);

// Update tokenValue property
await this.profileInfo.updateKnownProperty({ mergedArgs, property: 'tokenValue', value: '', setSecure: this.profileInfo.isSecured() });
await this.profileInfo.updateKnownProperty({ mergedArgs, property: 'tokenType', value: ''});
```
Here, the question is also how to correctly remove the `tokenValue` and `tokenType` properties. I tried setting the value of the properties to `undefined`, but that gives the following error:
```
Type 'undefined' is not assignable to type 'IProfArgValue'.ts(2322)
IProfInfoUpdatePropOpts.d.ts(39, 5): The expected type comes from property 'value' which is declared here on type 'IProfInfoUpdateKnownPropOpts'
```
Which makes sense based on the type declaration for IProfArgValue:
```typescript
export declare type IProfArgValue = string | number | boolean | string[] | object;
```

---

What I'm looking for is some direction on how to properly remove properties from the config file. As mentioned earlier, we can read the config json file in into the extension, remove the properties, and write it back to file. I'm not sure what is left behind, for example in the credential store...

In my opinion, it would be nice to have a method like `removeKnownProperty`, e.g.
```typescript
await this.profileInfo.removeKnownProperty({mergedArgs, property: 'tokenValue'});
```

**Comment by t1m0thyj** [comment, marked as answer]
Could you please try replacing the `updateKnownProperty` calls with the following?
```typescript
await this.profileInfo.updateKnownProperty({ mergedArgs, property: 'tokenValue', value: undefined as any, setSecure: false });
await this.profileInfo.updateKnownProperty({ mergedArgs, property: 'tokenType', value: undefined as any });
```

Two things to note here:
* The `undefined as any` should work to remove the property but of course the `as any` is not ideal. We should probably either update the type declaration of `IProfArgValue` to include `undefined`, or perhaps add a `removeKnownProperty` method as you suggested 🙂 
* The `setSecure` property must be set to false in order to remove an item from the "secure" array in the config JSON file.

**Reply by juleskreutzer** [reply]
Thanks @t1m0thyj for your answer. Setting the value to `undefined as any` does indeed solve my issue and remove the properties in the config JSON file. I also updated the setSecure to `false`

Regarding the `IProfArgValue` type, I think that adding an `undefined` is a good 'quick fix', but I'm wondering if it would be clear enough for others that setting the value to undefined will actually remove the property.

In my opinion, introducing a method like `removeKnownProperty` is self-explanatory and it's implementation could be as simple as calling the `updateKnownProperty` method with value `undefined as any`.

What is your opinion on this?

**Reply by t1m0thyj** [reply]
I agree with both suggestions - adding support for `undefined` as a value to pass to `updateKnownProperty`, and also adding a `removeKnownProperty` convenience method. We can track this here 🙂 https://github.com/zowe/imperative/issues/917
